### PR TITLE
Reduce scope and number of locks in Rust runtime

### PR DESCRIPTION
**Pull request by daviddrysdale**
# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

**Comment by daviddrysdale** [comment]
For #779

**Comment by daviddrysdale** [comment]
Yeah, there's definitely the chance of increasing lock contention with the need to use exclusive-`write` locks rather than shared-`read` locks when modifying.  One to remember when we get as far as performance tuning.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 1]
```diff
@@ -412,6 +405,7 @@ impl Runtime {
         // add itself again. Note that since that lock is currently held, the woken thread will add
         // itself to waiting_threads *after* we call clear below as we release the lock implicilty
         // on leaving this function.
+        let mut waiting_threads = channel.waiting_threads.lock().unwrap();
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/mod.rs:405]: (optional) Perhaps an explicit anonymous scope with a comment here may be useful to clarify what the scope of the lock is, especially if in the future some more instructions are added after `waiting_threads.clear()`.

Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs:405]: I'm [planning](https://github.com/project-oak/oak/issues/779#issuecomment-607170580) on killing the `waiting_threads` mutex soon anyhow &ndash; just need to untangle one place that takes the same lock twice (which is (apparently) OK for `RwLock::read` but goes wrong if one has to be converted to a `RwLock::write`).

**Pull request merged**